packages/webdriver/src/local.ts

```diff
@@ -0,0 +1,110 @@
+import 'chromedriver';
+import 'geckodriver';
+
+import { Operation, resource } from 'effection';
+import { spawn as spawnChildProcess } from '@effection/node';
+import { Atom } from '@bigtest/atom';
+import { once } from '@effection/events';
+
+import { fetch } from './fetch';
+import { findAvailablePortNumber } from './find-available-port-number';
+import { untilURLAvailable } from './until-url-available';
+import { WebDriver, Options } from './interfaces';
+
+type LocalDriverName = 'chromedriver' | 'geckodriver';
+
+/**
+ * Create a local `WebDriver` resource based on `driverName` (either 'geckodriver' or
+ * 'chromedriver').
+ *
+ * In order to function, the local executables for `Chrome` and
+ * `Firefox` must be separately installed onto the machine. So in CI
+ * for example, you would need to make sure that the browsers are
+ * present on your container.
+ *
+ * In order to keep things simple, and because the geckodriver only
+ * suopports a single Firefox session per driver process, we spawn a
+ * new process for each local driver.
+ */
+export function * Local(driverName: LocalDriverName, options: Options): Operation<WebDriver> {
+
+  let port: number = yield findAvailablePortNumber();
+  let driverURL = `http://localhost:${port}`;
+
+  let session: { sessionId: string };
+
+  let capabilities = new Atom(Capabilities);
+
+  if (options.headless) {
+    capabilities.slice<string[]>(['alwaysMatch', 'goog:chromeOptions', 'args'])
+      .over(args => args.concat(['--headless']))
+    capabilities.slice<string[]>(['alwaysMatch', 'moz:firefoxOptions', 'args'])
+      .over(args => args.concat(['--headless']))
+  }
+
+  let driver: WebDriver = {
+    *navigateTo(url: string): Operation<void> {
+      let response = yield fetch(`${driverURL}/session/${session.sessionId}/url`, {
+        method: 'post',
+        body: JSON.stringify({ url })
+      });
+
+      if (!response.ok) {
+        throw new Error(`GetError: GET ${url}\n${response.status} ${response.statusText}`);
+      }
+    }
+  };
+
+  yield resource(driver, function*() {
+    let child = yield spawnChildProcess(driverName, [`--port=${port}`]);
+
+    yield once(child, 'exit');
+  });
+
+
+  yield untilURLAvailable(`${driverURL}/status`, 1000);
+
+  let response: Response = yield fetch(`${driverURL}/session`, {
+    method: 'post',
+    body: JSON.stringify({
+      capabilities: capabilities.get()
+    })
+  });
+
+
+  if (!response.ok) {
+    let details: WDResponse;
+    try {
+      details = yield response.json();
+    } catch (e) { /* ok, no json details*/}
+    if (details && details.value && details.value) {
+      throw new Error(details.value.message);
+    } else {
+      throw new Error(`RequestError: ${response.status} ${response.statusText}`)
+    }
+  }
+
+  let json = yield response.json();
+
+  session = json.value;
```

Comment:
What's the point of assigning this variable?

Reply:
We need the `session.sessionId` in order to make further request to the Webdriver session.